path_dynamic_creds.go

```diff
@@ -0,0 +1,254 @@
+		roleNameRaw, ok := req.Secret.InternalData["name"]
+		if !ok {
+			return nil, fmt.Errorf("missing role name")
+		}
+
+		roleName := roleNameRaw.(string)
+		dRole, err := retrieveDynamicRole(ctx, req.Storage, roleName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve dynamic role: %w", err)
+		}
+		if dRole == nil {
+			return nil, fmt.Errorf("unable to renew: role does not exist")
+		}
+
+		// Update the default TTL & MaxTTL to the latest from the role in the event the role definition has changed
+		secret := req.Secret
+		secret.TTL = dRole.DefaultTTL
+		secret.MaxTTL = dRole.MaxTTL
+
+		resp := &logical.Response{
+			Secret: req.Secret,
+		}
+		return resp, nil
+	}
+}
+
+func (b *backend) secretCredsRevoke() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		config, err := readConfig(ctx, req.Storage)
+		if err != nil {
+			return nil, err
+		}
+		if config == nil {
+			return nil, fmt.Errorf("missing OpenLDAP configuration")
+		}
+
+		deletionTemplate, err := getString(req.Secret.InternalData, "deletion_ldif")
+		if err != nil {
+			return nil, fmt.Errorf("broken internal data: unable to retrieve deletion_ldif: %w", err)
+		}
+
+		_, err = b.executeLDIF(config.LDAP, deletionTemplate, req.Secret.InternalData["template_data"], true)
+		return nil, err
+	}
+}
+
+type usernameTemplateData struct {
+	DisplayName string
+	RoleName    string
+}
+
+const defaultUsernameTemplate = "v_{{.DisplayName}}_{{.RoleName}}_{{random 10}}_{{unix_time}}"
+
+func generateUsername(req *logical.Request, role *dynamicRole) (string, error) {
+	usernameTemplate := role.UsernameTemplate
+	if role.UsernameTemplate == "" {
+		usernameTemplate = defaultUsernameTemplate
+	}
+	tmpl, err := template.NewTemplate(
+		template.Template(usernameTemplate),
+	)
+	if err != nil {
+		return "", err
+	}
+	usernameData := usernameTemplateData{
+		DisplayName: req.DisplayName,
+		RoleName:    role.Name,
+	}
+	return tmpl.Generate(usernameData)
+}
+
+type dynamicTemplateData struct {
+	Username              string
+	Password              string
+	DisplayName           string
+	RoleName              string
+	IssueTime             string
+	IssueTimeSeconds      int64
+	ExpirationTime        string
+	ExpirationTimeSeconds int64
+}
+
+func applyTemplate(rawTemplate string, data dynamicTemplateData) (string, error) {
+	tmpl, err := template.NewTemplate(
+		template.Template(rawTemplate),
+	)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse template: %w", err)
+	}
+	str, err := tmpl.Generate(data)
+	if err != nil {
+		return "", fmt.Errorf("failed to execute template: %w", err)
+	}
+
+	return str, nil
+}
+
+func getString(m map[string]interface{}, key string) (string, error) {
+	if m == nil {
+		return "", fmt.Errorf("nil map")
+	}
+
+	val, exists := m[key]
+	if !exists {
+		return "", nil
```

Comment:
Should this be returning an error? I don't know if it's okay to pass in an empty string to `b.executeLDIF` in [here](https://github.com/hashicorp/vault-plugin-secrets-openldap/pull/21/files#diff-d5b65bfb287fcee080448ea1cbe9ba3f5af4f06b56b059bc2d0a59103f48a407R183).

Reply:
Good catch. I think this function should not return an error so it can distinguish between "not found" and "invalid type" or other error. I've added an additional check in `secretCredsRevoke` to check if the template is empty. This should only happen if there's a screw up on the developer's part since the `deletion_ldif` should already exist by this point.